python/ray/data/_internal/datasource/clickhouse_datasource.py

```diff
@@ -0,0 +1,201 @@
+                    elif operator == "not":
+                        filter_conditions.append(f"{key} IS NOT NULL")
+                elif isinstance(value, str):
+                    operator = validate_non_numeric_ops(key, operator)
+                    filter_conditions.append(f"{key} {ops[operator]} '{value}'")
+                elif isinstance(value, bool):
+                    operator = validate_non_numeric_ops(key, operator)
+                    filter_conditions.append(
+                        f"{key} {ops[operator]} {str(value).lower()}"
+                    )
+                elif isinstance(value, (int, float)):
+                    filter_conditions.append(f"{key} {ops[operator]} {value}")
+                else:
+                    logger.warning(
+                        f"Unsupported data type for filter on "
+                        f"'{key}': {type(value).__name__}"
+                    )
+
+            query += f" WHERE {' AND '.join(filter_conditions)}"
+
+        if self._order_by:
+            columns, desc = self._order_by
+            direction = " DESC" if desc else ""
+            if len(columns) == 1:
+                query += f" ORDER BY {columns[0]}{direction}"
+            elif len(columns) > 1:
+                columns_clause = ", ".join(columns)
+                query += f" ORDER BY ({columns_clause}){direction}"
+        return query
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        """
+        Estimate the in-memory data size for the query.
+
+        Returns:
+            Estimated in-memory data size in bytes, or
+             None if the estimation cannot be performed.
+        """
+        self._get_or_create_client()
+        estimate_query = f"SELECT SUM(byteSize(*)) AS total_size FROM ({self._query})"
+        try:
+            result = self._client.query(estimate_query)
+            if result and len(result.result_rows) > 0:
+                total_size = result.result_rows[0][0]
+                return int(total_size) if total_size is not None else None
+        except Exception as e:
+            logger.warning(f"Failed to estimate query size: {e}")
+        return None
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        """
+        Create read tasks for the ClickHouse query.
+
+        Args:
+            parallelism: The desired number of partitions to read the data into.
+
+        Returns:
+            A list of read tasks to be executed.
+        """
+        import pyarrow as pa
+
+        def _create_read_fn(partition_batches):
+            def _read_fn() -> Block:
+                yield pa.Table.from_batches(partition_batches)
+
+            return _read_fn
+
+        self._get_or_create_client()
+
+        # Fetch the fragments from the ClickHouse client
+        with self._client.query_arrow_stream(self._query) as stream:
+            record_batches = list(stream)  # Collect all record batches
```

Comment:
So actual reading of the data needs to be performed inside the read task (that's currently consolidated in `_read_fn`)

Reply:
I'd also recommend you to take a look at `SQLDatasource` to see how it could be structured in a way compatible with Ray read APIs

Reply:
Ah, I see what you mean. I'll make that change. Thank you for pointing that out!

Reply:
@alexeykudinkin Let me know if the changed I made resolve this!

**Evidence**
![Screenshot 2024-11-26 at 12 05 40 AM](https://github.com/user-attachments/assets/1968cfd1-cc83-42d7-a7e2-ce4f6b9cf639)

Reply:
Great!